aten/src/ATen/native/cuda/Shape.cu

```diff
@@ -0,0 +1,399 @@
+  * concatDim: dimension along which we are concatenating
+  * dimStride: the stride of the output tensor at the concatDim
+  *
+  * The most important assumption made is that the input tensors are contiguous.
+  */
+template <typename T, typename IndexType, int Dims>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(512)
+#endif
+__global__ void CatArrayBatchedCopy(
+    T* output,
+    CatArrInputTensor<T, IndexType>* inputs,
+    OutputTensorSizeStride<IndexType, CAT_ARRAY_MAX_INPUT_DIMS> os,
+    const int concatDim,
+    IndexType dimStride) {
+
+    IndexType tid = blockIdx.x * blockDim.x + threadIdx.x;
+    IndexType nElements = inputs[blockIdx.y].nElements;
+
+    if(tid >= nElements) return;
+    
+    T* data = inputs[blockIdx.y].input;
+    IndexType offset = inputs[blockIdx.y].offset;
+    IndexType dimSize = inputs[blockIdx.y].dimSize;
+    IndexType dataOffset = offset * dimStride;
+
+    IndexType stride = gridDim.x * blockDim.x;
+
+    while( tid < nElements){
+    IndexType elementOffset = CatArrIndexToOffset<IndexType, Dims>::compute(
+                  os.outputSize, os.outputStride, dimSize, concatDim, tid);
+    output[dataOffset + elementOffset] = data[tid];
+
+    tid += stride;
+    }
+}
+
+int64_t dim_size(const Tensor &self, int dim)
+{
+  TORCH_CHECK((dim >= 0) && (dim < self.dim()), 
+      "dimension ", dim, " out of range of tensor");
+  return at::native::size(self, dim);
+}
+
+int64_t dim_stride(const Tensor &self, int dim)
+{
+  TORCH_CHECK((dim >= 0) && (dim < self.dim()),
+      "dimension ", dim, " out of range");
+  return self.stride(dim);
+}
+
+void check_shape_except_dim(const Tensor &first, const Tensor &second,
+                            int dimension)
+{
+  int first_dims = first.dim();
+  int second_dims = second.dim();
+  TORCH_CHECK(first_dims == second_dims,
+      "Tensors must have same number of dimensions: got ", first_dims,
+      " and ", second_dims);
+  for (int dim = 0; dim < first_dims; dim++) {
+    if (dim == dimension) {
+      continue;
+    }
+    int64_t first_dim_size = at::native::size(first, dim);
+    int64_t second_dim_size = at::native::size(second, dim);
+    TORCH_CHECK(first_dim_size == second_dim_size,
+        "Sizes of tensors must match except in dimension ", dim, ". Got ",
+        static_cast<long long>(first_dim_size), " and ",
+        static_cast<long long>(second_dim_size));
+  }
+}
+
+template <typename scalar_t>
+void parallel_cat(Tensor &out, THCState *state, const TensorList &inputs,
+    int64_t dimension, int nDims) {
+  // First, let's set up our kernel parameters. We start with a raw pointer to
+  // the storage for the output Tensor.
+  scalar_t *data = out.data_ptr<scalar_t>();
+
+  // Kernel Parameter
+  size_t tensorMetadataSize =
+    sizeof(CatArrInputTensor<scalar_t, unsigned int>) * CAT_ARRAY_BATCH_SIZE;
+  // TODO: replacement for THCudaMalloc()
+  auto d_inputs = static_cast<CatArrInputTensor<scalar_t, unsigned int> *>(
```

Comment:
You can allocate ByteTensor of the necessary size and cast its data_ptr to the required type. You won't need to THCudaFree it in this case